packages/scripts/src/upgrade-tests/register-new-contract/1_Propose.ts

```diff
@@ -0,0 +1,333 @@
+// This script generates and submits the transaction to register in the Registry and add to the Finder an arbitrary new
+// contract in mainnet and layer 2 blockchains. It can be run on a local hardhat node fork of the mainnet or can be run
+// directly on the mainnet to execute the upgrade transactions.
+// To run this on the localhost first fork mainnet into a local hardhat node by running:
+// HARDHAT_CHAIN_ID=1 yarn hardhat node --fork https://mainnet.infura.io/v3/<YOUR-INFURA-KEY> --port 9545 --no-deploy
+// Then execute the script:
+// NODE_URL_10=<OPTIMISM-NODE-URL> \
+// \
+// NODE_URL_288=<OPTIMISM-NODE-URL> \
+// \
+// NODE_URL_137=<OPTIMISM-NODE-URL> \
+// \
+// NODE_URL_42161=<OPTIMISM-NODE-URL> \
+// \
+// yarn hardhat run ./src/upgrade-tests/register-new-contract/1_Propose.ts --network localhost
+
+const hre = require("hardhat");
+
+const { getAddress } = require("@uma/contracts-node");
+const { RegistryRolesEnum } = require("@uma/common");
+import { BigNumberish } from "@ethersproject/bignumber";
+import { BytesLike } from "@ethersproject/bytes";
+import { interfaceName } from "@uma/common";
+import {
+  ArbitrumParentMessenger,
+  Finder,
+  Governor,
+  GovernorHub,
+  GovernorRootTunnel,
+  Proposer,
+  Registry,
+} from "@uma/contracts-node/typechain/core/ethers";
+import { BaseContract, PopulatedTransaction, Signer } from "ethers";
+import { getContractInstance, getContractInstanceByUrl } from "../../utils/contracts";
+
+// PARAMETERS
+const proposerWallet = "0x2bAaA41d155ad8a4126184950B31F50A1513cE25";
+const newContractName = interfaceName.OptimisticAsserter;
+
+// CONSTANTS
+const POLYGON_CHAIN_ID = 137;
+const ARBITRUM_CHAIN_ID = 42161;
+
+// Env vars
+const NODE_URL_ENV = "NODE_URL_";
+
+export interface RelayProposal {
+  to: string;
+  value: BigNumberish;
+  data: BytesLike;
+  chainId: BigNumberish;
+}
+export interface RelayRecords {
+  block: number;
+  governorRootTunnel: RelayProposal[];
+  governorHub: RelayProposal[];
+}
+
+const relayGovernanceRootTunnelMessage = async (
```

Comment:
it feels like we now have a number of variants of this function spread over the scripts package (https://github.com/UMAprotocol/protocol/blob/6fddb23c8a62542084d6d72fdfab00213ffdcd8d/packages/scripts/src/admin-proposals/utils.js#L244 & https://github.com/UMAprotocol/protocol/blob/6fddb23c8a62542084d6d72fdfab00213ffdcd8d/packages/scripts/src/upgrade-tests/162/1_Propose.ts#L66 and others).

what do you think we should do about this repetition but slight variation in implementation details?

Reply:
in fact this function (and the following function) look to be exactly the same as that in 162 where we are dealing with ethers-based scripts. we could consider getting this PR in its current form, then, refactoring this `src/upgrade-tests` to have an `ethers` based sub-folder and then pull out these common functions into some common structure to remove this repetition.

Reply:
Totally agree @chrismaree,

I've done a refactor where I group all the relay functions in a [common file](https://github.com/UMAprotocol/protocol/blob/md0x/oa-proposals/packages/scripts/src/utils/relay.ts) 

Also I've grouped the relays for the GovernorHub l2 chains (Arbitrum, Optimism, Boba) so they are relayed in a single transaction instead of 4: [relayGovernanceHubMessages](https://github.com/UMAprotocol/protocol/blob/md0x/oa-proposals/packages/scripts/src/utils/relay.ts#L32)

If you think this covers your suggestions, after merging this PR we could refactor/delete [162 folder](https://github.com/UMAprotocol/protocol/tree/master/packages/scripts/src/upgrade-tests/162) as [register-new-contract](https://github.com/UMAprotocol/protocol/tree/md0x/oa-proposals/packages/scripts/src/upgrade-tests/register-new-contract) as a generic way to register any new contract in mainnet + l2s

Reply:
yup! that refactor looks much better. we can move stuff around for the 162 folder once this is in. good call. I'm going to look more closely into the actual governance transactions produced here now and will report back.